File: src/jastudio_src/jastudio/note/vocabulary/vocabnote_userfields.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from jastudio.note.note_constants import NoteFields
from jastudio.note.notefields.caching_mutable_string_field import CachingMutableStringField
from jastudio.note.notefields.mutable_string_field import MutableStringField

if TYPE_CHECKING:
    from jaslib.sysutils.weak_ref import WeakRef
    from jastudio.note.vocabulary.vocabnote import VocabNote

class VocabNoteUserfields(Slots):
    def __init__(self, vocab: WeakRef[VocabNote]) -> None:
        self._vocab: WeakRef[VocabNote] = vocab
        self.answer: CachingMutableStringField = CachingMutableStringField(vocab, NoteFields.Vocab.user_answer)

    @property
    def mnemonic(self) -> MutableStringField: return MutableStringField(self._vocab, NoteFields.Vocab.user_mnemonic)
    @property
    def explanation(self) -> MutableStringField: return MutableStringField(self._vocab, NoteFields.Vocab.user_explanation)
    @property
    def explanation_long(self) -> MutableStringField: return MutableStringField(self._vocab, NoteFields.Vocab.user_explanation_long)

    @override
    def __repr__(self) -> str: return f"Answer: {self.answer.value}, Mnemonic: {self.mnemonic.value}, Explanation: {self.explanation.value}, Explanation Long: {self.explanation_long.value}"

File: src/src_dotnet/JAStudio.Core/Note/Vocabulary/VocabNoteUserFields.cs
using JAStudio.Core.Note.CorpusData;
using JAStudio.Core.Note.NoteFields;

namespace JAStudio.Core.Note.Vocabulary;

public class VocabNoteUserFields
{
   public WritableStringValue Answer { get; }
   public WritableStringValue Mnemonic { get; }
   public WritableStringValue Explanation { get; }
   public WritableStringValue ExplanationLong { get; }

   public VocabNoteUserFields(VocabData? data, NoteGuard guard)
   {
      Answer = new WritableStringValue(data?.UserAnswer ?? string.Empty, guard);
      Mnemonic = new WritableStringValue(data?.UserMnemonic ?? string.Empty, guard);
      Explanation = new WritableStringValue(data?.UserExplanation ?? string.Empty, guard);
      ExplanationLong = new WritableStringValue(data?.UserExplanationLong ?? string.Empty, guard);
   }

   public override string ToString() =>
      $"Answer: {Answer.Value}, Mnemonic: {Mnemonic.Value}, Explanation: {Explanation.Value}, Explanation Long: {ExplanationLong.Value}";
}
